### ISSUE TITLE
F# interactive command history is off by one

### ISSUE THREAD

**Issue by jwosty** [mono/monodevelop#4584]
I don't see this issue filed elsewhere, but this has been present for a while for me. I'm running VS for Mac on OSX.

In F# interactive, when you press the up-arrow, it wrongly starts at the *second* to last command you entered instead of the _last_ command. To reproduce:

<s>1) Fire up VS for Mac (I assume the problem is also present in MD/XS).</s>
<s>2) In F# interactive, enter `printfn "1st line"`</s>
<s>3) Enter `printfn "2nd line"`</s>
<s>4) Press the up-arrow once, and notice that it jumps to `printfn "1st line`".</s>

_This actually isn't correct; see [these steps](https://github.com/mono/monodevelop/issues/4584#issuecomment-384816227) instead._

The history is all there -- if you then press the down arrow once, it indeed fills in `printfn "2nd line"`.

> VS bug [#601374](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/601374)

**Comment by jwosty** [comment]
So I've been investigating further, and reproduction is actually a little more tricky than I thought. It's only off by one after something happens, and it stays this way for the rest of the session (even after you reset the session). It always seems to involve some amount of sending things to interactive. After some playing around, here's a set of minimum steps that definitely ruin a session:

1. Directly enter `1;;`
2. Directly enter `2;;`
3. Highlight + send the following to interactive (cmd+enter) from a script file:
```fsharp
let f x =
    printfn "a"
    printfn "b"
    printfn "x = %s" x
```
4. Highlight + send the following to interactive (cmd+enter) from a script file:
```fsharp
f "foo"
```
5. Now, observe the the up-arrow fills in `1;;`, when it *should* be `2;;`. Session now seems to be permanently off by one on further invocations (e.g. try directly entering another line and pressing up arrow)

**Comment by nosami** [comment]
Can confirm. I've been aware of this behaviour for a while but it was quite low down on the list of priorities. I'll take a look into it now.

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-17] Comment by **xamarinc**



Fixed in version 8.6.0.802 (master)



Pull Request #4860 merged by: Jason Imison  
 Author: mono  
 Commit: 470ba8405c8e756271625f6581bb293082b54817 (xamarin/monodevelop)
